python_modules/dagster-graphql/dagster_graphql/implementation/loader.py

```diff
@@ -277,3 +283,77 @@ def get_latest_materialization_for_asset_key(self, asset_key: AssetKey) -> Event
     def _fetch(self):
         self._fetched = True
         self._materializations = self._instance.get_latest_materialization_events(self._asset_keys)
+
+
+class BatchAssetDependencyLoader:
+    def __init__(self, context: WorkspaceRequestContext):
+        self._context = context
+        self._fetched = False
+        self._sink_assets: Dict[AssetKey, ExternalAssetNode] = {}
+        self._external_asset_deps: Dict[
+            Tuple[str, str], Dict[AssetKey, Sequence[ExternalAssetDependedBy]]
+        ] = (
+            {}
+        )  # nested dict that maps dependedby assets by asset key by location tuple (repo_location.name, repo_name)
+
+    def _build_cross_repo_deps(self):
+        depended_by_assets_by_source_asset: Dict[AssetKey, Sequence[ExternalAssetDependedBy]] = {}
+
+        map_defined_asset_to_location: Dict[
+            AssetKey, Tuple[str, str]
+        ] = {}  # key is asset key, value is tuple (location_name, repo_name)
+        for location in self._context.repository_locations:
+            repositories = location.get_repositories()
+            for repo_name, external_repo in repositories.items():
+                asset_nodes = external_repo.get_external_asset_nodes()
+                for asset_node in asset_nodes:
+                    if not asset_node.op_name:  # is source asset
+                        if asset_node.asset_key not in depended_by_assets_by_source_asset:
+                            depended_by_assets_by_source_asset[asset_node.asset_key] = []
+                        depended_by_assets_by_source_asset[asset_node.asset_key].extend(
+                            asset_node.depended_by
+                        )
+                    else:
+                        map_defined_asset_to_location[asset_node.asset_key] = (
+                            location.name,
+                            repo_name,
+                        )
+
+        for source_asset, depended_by_assets in depended_by_assets_by_source_asset.items():
+            asset_def_location = map_defined_asset_to_location.get(source_asset, None)
+            if asset_def_location:  # source asset is defined as asset in another repository
+                if asset_def_location not in self._external_asset_deps:
+                    self._external_asset_deps[asset_def_location] = {}
+                if source_asset not in self._external_asset_deps[asset_def_location]:
+                    self._external_asset_deps[asset_def_location][source_asset] = []
+                self._external_asset_deps[asset_def_location][source_asset].extend(
+                    depended_by_assets
+                )
+                for asset in depended_by_assets:
+                    self._sink_assets[asset.downstream_asset_key] = ExternalAssetNode(
```

Comment:
As discussed on zoom, would it make sense to just use the `ExternalAssetNode` from the downstream repository instead of constructing a new one?

Reply:
Ah actually, looks like source assets defined as `ExternalAssetNodes` contain different information than derived assets ([file here](https://github.com/dagster-io/dagster/blob/2e6eabd0f54e318de3cab7ef4712c13e2ad4b21e/python_modules/dagster/dagster/core/host_representation/external_data.py#L857)). Dagit will display any `ExternalAssetNode` with null `job_names` param as a foreign asset, so using the derived asset's `ExternalAssetNode` will display in Dagit as a regular node instead of an external node.

Reply:
What exactly does it mean to "display as a foreign asset?" Show a link instead of a box when inside the asset graph view? I thought that was based on what assets are selected?

Reply:
If I'm understanding correctly, that means it's the frontend, not the backend code here, that determines whether an asset displays as a regular node or an external node? If that's the case, do we need to hold back information here?

Reply:
The frontend reads `job_names` from the `ExternalAssetNode` object. If `job_names=[]`, then the asset displays as an external node. Otherwise, it displays as a regular node

Reply:
Ah ok, so the backend does play a part in it. That strikes me as a little strange, but out of the scope of this PR. So I'm on board with this part of the implementation.